app/src/types/api.js

```diff
@@ -6,37 +6,99 @@
  * Contains io-ts models for the API response types.
  */
 
-import * as t from "io-ts/lib/index";
-
-export const GetProfileOKResponseModel = t.partial({
-  preferredLanguages: t.readonlyArray(t.string),
-  email: t.string,
-  isInboxEnabled: t.boolean,
-  version: t.number
-});
-
-export const UpsertProfileOKResponseModel = t.partial({
-  preferredLanguages: t.readonlyArray(t.string),
-  email: t.string,
-  isInboxEnabled: t.boolean,
-  version: t.number
-});
-
-export const ProblemJsonModel = t.partial({
-  type: t.string,
-  title: t.string,
-  status: t.number,
-  detail: t.string,
-  instance: t.string
-});
-
-export const GetMessagesByUserOKResponseModel = t.partial({
-  pageSize: t.number,
-  next: t.string,
-  items: t.readonlyArray(t.any)
-});
-
-export const MessageResponseModel = t.partial({
-  message: t.any,
-  notification: t.any
-});
+import t from "flow-runtime";
+import {
+  GetMessagesByUserOKResponse,
+  GetProfileOKResponse,
+  MessageResponse,
+  UpsertProfileOKResponse
+} from "../api/models";
+import { left, right } from "fp-ts/lib/Either";
+
+const GetProfileOKResponseModel = t.object(
+  t.property("preferredLanguages", t.array(t.string()), true),
+  t.property("email", t.string(), true),
+  t.property("isInboxEnabled", t.boolean(), true),
+  t.property("version", t.number(), true)
+);
+
+const UpsertProfileOKResponseModel = t.object(
+  t.property("preferredLanguages", t.array(t.string()), true),
+  t.property("email", t.string(), true),
+  t.property("isInboxEnabled", t.boolean(), true),
+  t.property("version", t.number(), true)
+);
+
+const ProblemJsonModel = t.object(
+  t.property("type", t.string(), true),
+  t.property("title", t.string(), true),
+  t.property("status", t.number(), true),
+  t.property("detail", t.string(), true),
+  t.property("instance", t.string(), true)
+);
+
+const GetMessagesByUserOKResponseModel = t.object(
+  t.property("pageSize", t.number(), true),
+  t.property("next", t.string(), true),
+  t.property("items", t.array(t.any()), true)
+);
+
+const MessageResponseModel = t.object(
+  t.property("message", t.any(), true),
+  t.property("notification", t.any(), true)
+);
+
+/**
+ * Validates on object against the ProblemJsonModel data type.
+ *
+ * @param value
+ * @param res
+ */
+export function validateProblemJson(value: any, res: express$Response) {
+  t.validate(ProblemJsonModel, value).fold(
+    () => {
+      res.status(500).json({
+        // If we reach this point something very bad has happened.
+        message: "Unrecoverable error."
+      });
+    },
+    error => {
+      res.status(error.status).json({
+        // Forward the error received from the API.
+        message: error.detail
+      });
+    }
+  );
+}
+
+export function validateGetProfileOKResponseModel(
```

Comment:
I guess these could be generalized (`validateResponse(response, type)`)

Reply:
side note,  it looks like t.validate has a fold method so I guess this is already foldable, is that correct ?
because in case you can just say


`validateResponse(r, t) = t.validate(r, t).fold(left, right);`